larpandoracontent/LArVertex/TrainedVertexSelectionAlgorithm.cc

```diff
@@ -582,6 +685,96 @@ const pandora::Vertex *TrainedVertexSelectionAlgorithm::ProduceTrainingExamples(
 
 //------------------------------------------------------------------------------------------------------------------------------------------
 
+void TrainedVertexSelectionAlgorithm::GetSharedFeatures(
+    const Vertex *const pVertex1, const Vertex *const pVertex2, const KDTreeMap &kdTreeMap, float &separation, float &axisHits) const
+{
+    separation = (pVertex1->GetPosition() - pVertex2->GetPosition()).GetMagnitude();
+
+    this->IncrementSharedAxisValues(LArGeometryHelper::ProjectPosition(this->GetPandora(), pVertex1->GetPosition(), TPC_VIEW_U),
+        LArGeometryHelper::ProjectPosition(this->GetPandora(), pVertex2->GetPosition(), TPC_VIEW_U), kdTreeMap.at(TPC_VIEW_U), axisHits);
+
+    this->IncrementSharedAxisValues(LArGeometryHelper::ProjectPosition(this->GetPandora(), pVertex1->GetPosition(), TPC_VIEW_V),
+        LArGeometryHelper::ProjectPosition(this->GetPandora(), pVertex2->GetPosition(), TPC_VIEW_V), kdTreeMap.at(TPC_VIEW_V), axisHits);
+
+    this->IncrementSharedAxisValues(LArGeometryHelper::ProjectPosition(this->GetPandora(), pVertex1->GetPosition(), TPC_VIEW_W),
+        LArGeometryHelper::ProjectPosition(this->GetPandora(), pVertex2->GetPosition(), TPC_VIEW_W), kdTreeMap.at(TPC_VIEW_W), axisHits);
+
+    if (!separation)
+        axisHits = 0;
+
+    axisHits /= separation;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void TrainedVertexSelectionAlgorithm::IncrementSharedAxisValues(
+    const CartesianVector pos1, const CartesianVector pos2, HitKDTree2D &kdTree, float &axisHits) const
+{
+    if (pos1 == pos2)
+        return;
+
+    // Define the axis and perpendicular directions
+    const CartesianVector unitAxis = (pos1 - pos2).GetUnitVector();
+    const CartesianVector unitPerp(unitAxis.GetZ(), 0, -unitAxis.GetX());
+
+    // Define the corners of the search box
+    const CartesianVector point1 = pos1 + unitPerp;
+    const CartesianVector point2 = pos1 - unitPerp;
+    const CartesianVector point3 = pos2 + unitPerp;
+    const CartesianVector point4 = pos2 - unitPerp;
+
+    // Find the total coordinate span these points cover
+    float xMin(std::numeric_limits<float>::max());
+    float xMax(-std::numeric_limits<float>::max());
+    float zMin(std::numeric_limits<float>::max());
+    float zMax(-std::numeric_limits<float>::max());
+
+    this->UpdateBoxEdges(point1, xMin, xMax, zMin, zMax);
+    this->UpdateBoxEdges(point2, xMin, xMax, zMin, zMax);
+    this->UpdateBoxEdges(point3, xMin, xMax, zMin, zMax);
+    this->UpdateBoxEdges(point4, xMin, xMax, zMin, zMax);
```

Comment:
You'd want to double-check this, but I think you can get the same result by simply defining

```
const float xMin{std::min({point1.GetX(), point2.GetX(), point3.GetX(), point4.GetX()})}
```
And equivalently for the other minima and maxima
(note the curly braces that allow std::min to compare more than two values).

Reply:
I'll take a look, if so this is much neater thank you!